### PR TITLE
Add functions to allocate zeroed `Arc` and `Rc`

### PR DESCRIPTION
**Pull request by a1phyr**
These require rustc 1.82

### CODE REVIEW

File: src/allocation.rs [hunk 1]
```diff
@@ -149,6 +149,42 @@ pub fn zeroed_slice_box<T: Zeroable>(length: usize) -> Box<[T]> {
   try_zeroed_slice_box(length).unwrap()
 }
 
+/// Allocates a `Arc<T>` with all contents being zeroed out.
+pub fn zeroed_arc<T: Zeroable>() -> Arc<T> {
```
Comment (zachs18) [inline src/allocation.rs:153]:
```suggestion
/// Allocates a `Arc<T>` with all contents being zeroed out.
#[cfg(target_has_atomic = "ptr")]
pub fn zeroed_arc<T: Zeroable>() -> Arc<T> {
```

File: src/allocation.rs [hunk 2]
```diff
@@ -149,6 +149,42 @@ pub fn zeroed_slice_box<T: Zeroable>(length: usize) -> Box<[T]> {
...
+  unsafe {
+    let arc = Arc::new_uninit();
+    core::ptr::write_bytes(Arc::as_ptr(&arc) as *mut T, 0, 1);
+    arc.assume_init()
+  }
+}
+
+/// Allocates a `Arc<[T]>` with all contents being zeroed out.
+pub fn zeroed_arc_slice<T: Zeroable>(length: usize) -> Arc<[T]> {
```
Comment (zachs18) [inline src/allocation.rs:162]:
```suggestion
/// Allocates a `Arc<[T]>` with all contents being zeroed out.
#[cfg(target_has_atomic = "ptr")]
pub fn zeroed_arc_slice<T: Zeroable>(length: usize) -> Arc<[T]> {
```

**Review by zachs18 — CHANGES_REQUESTED** [review]
so that this will still compile on targets without `Arc`

**Comment by zachs18** [comment]
Also: It is currently up in the air if it is valid to write to the pointer returned from `{Arc,Rc}::as_ptr`: https://github.com/rust-lang/rust/issues/87862 https://github.com/rust-lang/rust/pull/104337 , so ideally these would use `{Arc,Rc}::get_mut(&mut arc).unwrap()` at least until that is clarified[^1].

[^1]: The compiler can optimize out the `unwrap` panic branch for `Rc` but not `Arc`, but it should still never be possible to panic at runtime since the `Arc` is newly allocated

File: src/allocation.rs (hunk 1 above)
Comment (a1phyr) [reply src/allocation.rs]: I also forgot `cfg(feature = "alloc_uninit")`, surprised that the CI didn't catch it

**Comment by a1phyr** [comment]
As you wish. I updated the code to use `write_zeroes` and `fill_zeroes`, it is cleaner like that if we use `get_mut`

**Review by Lokathor — APPROVED** [review]

**Comment by Lokathor** [comment]
released in bytemuck-1.20

**Pull request merged**